### DISCUSSION TITLE
@prop type for `keyof typeof MyObj`

### DISCUSSION THREAD

**Discussion by RemyMachado** [typegoose/typegoose#767] [Q&A] [answered]
I'm getting a type mixed warning with the following:

```typescript
// Typegoose model

    @prop({ required: true })
    breakpoint!: keyof typeof breakpoints
```

```typescript
    breakpoints = {
        fourK: 3860,
        qHD: 2560,
        ...
    }
```

I can't find how to define it properly.

Any advice would be greatly appreciated.

**Comment by hasezoey** [comment, marked as answer]
Reflection cannot type objects, so the only (proper) way is to outsource objects to their own class

instead of:
```ts
class C1 {
  @prop()
  public someobj: {
    someprop: string;
  }
}
```
do:
```ts
class Lower {
  @prop()
  public someprop: string;
}
class C1 {
  @prop()
  public someobj: Lower;
}
```

also look at the examples in the [Quick Start Guide](https://typegoose.github.io/typegoose/docs/guides/quick-start-guide)

also see [Nesting Classes](https://typegoose.github.io/typegoose/docs/guides/nesting-classes)

PS: if you want the keys of an object, you either have to manually type them out, or generate them on runtime:
```ts
const someobj = { key1: 100 };

class C1 {
  @prop({ enum: Object.keys(someobj) })
  public keyof: string;
}
```

**Reply by RemyMachado** [reply]
Thank you for the prompt response

Hmmm, I'm not sure I want to translate it to a class.

I'm gonna leave it as a mixed type.
Just to be sure to understand the meaning of "mixed type" -> it's allowing any type, similarly to putting TS `any`?

My GQL-codegen setup is outputting `breakpoint: Scalars['String'];` it's not as bad as I would expect.

**Reply by hasezoey** [reply]
>  Just to be sure to understand the meaning of "mixed type" -> it's allowing any type, similarly to putting TS any?

yes, meaning there is literally no validation or transformation of the values inside (and may need manual marking if modified (`doc.markModified`))

---

i have also updated my answer to include the actual `keyof` at runtime example (because i didnt read the question example code correctly)

**Reply by RemyMachado** [reply]
> const someobj = { key1: 100 };
>
> class C1 {
>   @prop({ enum: Object.keys(someobj) })
>   public keyof: string;
> }

No more mixed type with your updated example.

Thank you very much!

**Reply by hasezoey** [reply]
just as a note, that way only works if the object is defined at the point of where the `@prop` line is, i dont know if `enum` supports the generator / deferred function "syntax"
(this note is for issues like defining the object after the class or circular imports)
